File: models/Block.js
/// <reference path="../type_declarations/index.d.ts" />
var _ = require('lodash');
var async = require('async');
var db = require('../db');
var tree = require('../lib/tree');
var Block = (function () {
    function Block() {
    }
    /**
    shapeTree takes a list of blocks and returns a subset of that list, but where
    all given blocks can be found within the tree structure of each block in that list.
    */
    Block.shapeTree = function (blocks) {
        var block_hash = _.object(blocks.map(function (block) {
            block.children = [];
            return [block.id, block];
        }));
        var root_blocks = [];
        blocks.forEach(function (block) {
            if (block.parent_block_id) {
                // block_hash and root blocks contents are linked by reference, so order doesn't matter here
                block_hash[block.parent_block_id].children.push(block);
            }
            else {
                // blocks with no parent_block_id are added to the root list
                root_blocks.push(block);
            }
        });
        // I think these should be sorted already? maybe not
        return _.sortBy(root_blocks, 'view_order');
    };
    /**
    Searches for the first block that:
    1) is in the same experiment
    2) has the lowest view_order
    3) that's greater than the current block's view_order
  
    callback(Error)                  // unanticipated error
    callback(null, block_id: Number) // no error
    callback(null, null)             // no more blocks in experiment
    */
    Block.nextBlockId = function (experiment_id, block_id, participant_id, callback) {
        // this query is fast in the database (like 3ms) but can take much longer in javascript (like 25ms)
        // if we're requesting the whole block row, with its bulky context, so we
        // only select the columns we need
        db.query("SELECT blocks.id, blocks.template_id, blocks.template_id, blocks.randomize, blocks.quota,\n        blocks.parent_block_id, blocks.view_order, participant_responses.created AS completed FROM blocks\n      LEFT OUTER JOIN (SELECT DISTINCT ON (block_id) * FROM responses WHERE responses.participant_id = $1)\n        AS participant_responses ON participant_responses.block_id = blocks.id\n      WHERE blocks.experiment_id = $2\n      ORDER BY view_order ASC\n    ", [participant_id, experiment_id], function (err, blocks) {
            if (err)
                return callback(err);
            // should id = null?
            var root = {
                id: null,
                randomize: false,
                children: Block.shapeTree(blocks),
            };
            function findBlock(id) {
                return tree.recursiveFind([root], function (block) { return block.id === id; });
            }
            // we need to get the parent_block_id of the most_recent block
            var most_recent_block = findBlock(block_id);
            if (most_recent_block === undefined) {
                // not funny. that block doesn't exist in this experiment.
                return callback(new Error("Block with id=" + block_id + " cannot be found."));
            }
            // var current_parent_block_id = most_recent_block.parent_block_id;
            // 1. get the path of block ids (only ids, since we mutate the tree when
            //    pruning) from the parent of the most recent block to the root.
            /**
            Returns a list from nearest to root, e.g.:
                [parent_of_most_recent, grandparent_of_most_recent, ..., null]
            */
            function findPath(block_id) {
                var block = findBlock(block_id);
                if (block.parent_block_id) {
                    return [block].concat(findPath(block.parent_block_id));
                }
                return [block, root];
            }
            var current_path = findPath(most_recent_block.parent_block_id);
            // 2. travel upward until we find a parent that has incomplete children,
            //    marking blocks for completion as needed
            async.whilst(function () {
                var next_parent_block = current_path[0];
                if (next_parent_block === root) {
                    return false;
                }
                var completed_children = next_parent_block.children.filter(function (block) { return !!block.completed; });
                if (completed_children.length == next_parent_block.children.length) {
                    return true;
                }
                // check if the parent block's limit ("quota") has been reached
                if (next_parent_block.quota && completed_children.length >= next_parent_block.quota) {
                    return true;
                }
                return false;
            }, function (callback) {
                // okay, we're done with that block, shift it off
                var block = current_path.shift();
                // mark it completed locally, so that parent blocks will see it's been completed
                block.completed = new Date(); // doesn't really matter what date, actually
                // and then persist the completion, too
                db.InsertOne('responses')
                    .set({
                    participant_id: participant_id,
                    block_id: block.id,
                })
                    .execute(callback);
            }, function (error) {
                if (error)
                    return callback(error);
                // ok, current_path is either empty or current_path[0] is the next block we should do
                var next_parent_block = current_path[0];
                // 3. recurse downward and randomize as needed. this is async since we
                // may need to query past responses when randomizing
                // var next_parent_block = next_parent_block;
                // var next_block_id = findNextBlock(next_parent_block);
                // async.whilst(() => {
                (function next(next_parent_block) {
                    nextChildBlock(next_parent_block, function (error, next_block) {
                        if (error)
                            return callback(error);
                        // if there are no more blocks to do, next_block will be null
                        if (next_block === null) {
                            return callback(null, null);
                        }
                        // only return blocks with a template_id field
                        if (next_block.template_id) {
                            callback(null, next_block.id);
                        }
                        else {
                            next(next_block);
                        }
                    });
                })(next_parent_block);
            });
        });
    };
    return Block;
})();
function nextChildBlock(parent_block, callback) {
    var incomplete_child_blocks = parent_block.children.filter(function (block) { return !block.completed; });
    // if next_parent_block was `root` all the children might be completed
    if (incomplete_child_blocks.length === 0) {
        // so, we've completed all there is to complete!
        return callback(null, null);
    }
    if (parent_block.randomize) {
        if (parent_block.quota !== null) {
            var incomplete_child_block_ids = incomplete_child_blocks.map(function (block) { return block.id; });
            // find the next candidate that has the fewest completions
            db.Select('responses')
                .add('block_id, COUNT(id)::int')
                .where('block_id = ANY(?)', incomplete_child_block_ids)
                .groupBy('block_id')
                .execute(function (error, counts) {
                if (error)
                    return callback(error);
                var block_counts = _.zipObject(counts.map(function (count) { return [count.block_id, count.count]; }));
                // it's possible there are no responses for any of the candidate blocks
                incomplete_child_blocks.forEach(function (block) {
                    block.responses = block_counts[block.id] || 0;
                });
                var minimum_count = _.min(incomplete_child_blocks.map(function (block) { return block.responses; }));
                var candidate_child_blocks = incomplete_child_blocks.filter(function (block) { return block.responses == minimum_count; });
                //
                var next_block = _.sample(candidate_child_blocks);
                callback(null, next_block);
            });
        }
        else {
            setImmediate(function () {
                // _.sample called with a single argument returns a single item, not a list
                var next_block = _.sample(incomplete_child_blocks);
                callback(null, next_block);
            });
        }
    }
    else {
        setImmediate(function () {
            var next_block = _.first(incomplete_child_blocks);
            callback(null, next_block);
        });
    }
}
module.exports = Block;

File: src/formious/db/block.clj
(ns formious.db.block
  (:require [clojure.walk :as walk]
            [clojure.tools.logging :as log]
            [formious.db.response :as Response]
            [formious.util :refer [as-long]]
            [formious.db :as db]))

(defn all
  [experiment_id]
  (db/query ["SELECT * FROM block WHERE experiment_id = ? ORDER BY view_order ASC" (as-long experiment_id)]))

(defn next-in-experiment
  [experiment_id]
  (first (db/query ["SELECT * FROM block
                     WHERE experiment_id = ?
                     ORDER BY view_order ASC
                     LIMIT 1" (as-long experiment_id)])))

(defn find-by-id
  [experiment_id id]
  (first (db/query ["SELECT * FROM block
                     WHERE experiment_id = ?
                     AND id = ?" (as-long experiment_id) (as-long id)])))

(defn insert!
  [row]
  (db/insert! "block" row))

(defn update!
  [id experiment_id set-map]
  (db/update! "block" set-map ["id = ? AND experiment_id = ?" (as-long id) (as-long experiment_id)]))

(defrecord TreeNode [children])
(defrecord BranchNode [children value])
(defrecord RootNode [children])

(defn- get-node-value-id
  [node]
  (get-in node [:value :id]))

(defn- buildBranchNode
  "Take a root block and build it out into a structured tree,
  using children-map to (recursively) find children.
  * `parent-block` is the starting block
  * `children-map` is a mapping from block ID to that block's children (if any)
  Returns a structured tree"
  [parent-block children-map]
  (let [children (get children-map (:id parent-block) [])
        children-nodes (map #(buildBranchNode % children-map) children)]
    (BranchNode. children parent-block)))

(defn buildTree
  ; (blocks: Seq[Block]): RootNode[Block]
  [blocks]
  (let [children-map (group-by :parent_block_id blocks)
        ; blocks with no parent_block_id are added to the root list
        roots (get children-map nil)
        branches (map #(buildBranchNode % children-map) roots)]
    ; I think branches should be sorted already? maybe not
    (RootNode. (sort-by #(get-in % [:value :view_order]) branches))))

; Generate mapping from block.id to total number of responses for that block
(defn- fetchBlockResponseCounts
  ; blockIds: Seq[Int]): Map[Int, Long]
  [block-ids]
  (let [rows (db/query ["SELECT block_id, COUNT(id) FROM response
                         WHERE block_id = ANY(?)
                         GROUP BY block_id" block-ids])]
    ; TODO: it's possible there are no responses for any of the candidate blocks, so we supply a default
    ; .withDefaultValue(0L)
    (into {} (map (juxt :block_id :count) rows))))

;
; Find the next block that needs to be done. If randomize is false, just take the first of the given blocks.
; If randomize is true, and there is a quota, pick a less-completed (by previous participants) block, otherwise
; just pick a random block.
;
; @param blockNodes A non-empty sequence of Block nodes, which have been already determined to be incomplete.
;                   We have to handle block nodes rather than just Blocks, since we might need to recurse down
;                   to the children, if the selected block is a container (parent).
; @return One of the given blocks
;
(defn sampleBlockNode
  ; (blockNodes: Seq[BranchNode[Block]], randomize: Boolean, quotaOption: Option[Int]): BranchNode[Block]
  [blockNodes randomize quotaOption]
  ; TODO: handle case where quota is true, but randomize is false (does that make any sense, though?)
  (if randomize
    (if quotaOption
      (let [blockResponseCount (fetchBlockResponseCounts (map get-node-value-id blockNodes))
            ; sort by which block has the fewest completions
            sortedBlockNodes (sort-by #(blockResponseCount get-node-value-id) blockNodes)
            ; TODO: randomly sample weighted by missing completions, maybe?
            minimumResponseCount (-> sortedBlockNodes first :value :id blockResponseCount)
            candidateBlockNodes (take-while #(= (blockResponseCount get-node-value-id) minimumResponseCount) sortedBlockNodes)]
        (rand-nth candidateBlockNodes))
      (rand-nth blockNodes))
    (first blockNodes)))

(defn blockNodeQuota
  ; (blockNode: TreeNode[Block]): Int = blockNode match {
  [{:keys [children value]}]
  (if value
    ; case BranchNode(children, value) =>
    (if-let [quota (:quota value)]
      quota
      ; hopefully quota ≤ children.size, because otherwise it wouldn't make sense
      (count children))
    (count children)))

(defn nextPhysicalBlock
  "For the given blockNode, which should be pre-determined to have at least one incomplete child block,
  determine the next block needing to be completed.
  * `parentBlockNode` should have in its descendents somewhere an incomplete block
  * `completedBlocks` is a set of completed block IDs.
  Returns a block structure."
  [parentBlockNode completedBlocks]
  (let [incompleteBlockNodes (->> parentBlockNode
                                  :children
                                  (remove #(get completedBlocks (get-node-value-id %))))
        [randomize quota] (if-let [value (:value parentBlockNode)]
                            [(:randomize value) (:quota value)] ; branch node
                            [false nil]) ; root node
        childBlockNode (sampleBlockNode incompleteBlockNodes randomize quota)]
    ; recurse downward and randomize as needed.
    ; this is recursive since we may need to query past responses when randomizing
    (if (get-in childBlockNode [:value :template_id])
      ; only return blocks with a template_id field
      childBlockNode
      (nextPhysicalBlock childBlockNode completedBlocks))))

(defn nextBlock
  "Searches for the first block that:
  1) is in the same experiment
  2) has the lowest view_order
  3) that's greater than the current block's view_order
  Returns the next block to be completed,
  or nil if there are no more incomplete blocks in experiment."
  ; TODO: maybe select only the columns we need? I.e., exclude the bulky context
  [experiment_id block_id participant_id]
  (let [blocks    (all experiment_id)
        responses    (Response/whereParticipant participant_id)
        ; set of completed block ids for this participant
        originalCompletedBlocks (set (map get-node-value-id responses))
        rootNode    (buildTree blocks)
        currentBlockNodeOption []];   (rootNode.findFirst(_.id == block_id)
        ; if (currentBlockNodeOption.isEmpty) {
        ;   ; not funny. that block doesn't exist in this experiment.
        ;   logger.debug(s"Block with id=$block_id cannot be found.")
        ; }
    ; we need to get the parent_block_id of the most_recent block
    (mapcat identity currentBlockNodeOption)))
      ; currentBlockNode =>
      ; ; 1. get the path of blocks from the parent of the current block to the root.
      ; val parents = rootNode.parentPath(currentBlockNode)
      ; ; 2. travel upward until we find a parent that has incomplete children,
      ; ;    marking blocks as complete as needed
      ; val (nextParentBlockNodeOption, allCompletedBlocks) = parents.foldLeft((Option.empty[TreeNode[Block]], originalCompletedBlocks)) {
      ;   case ((Some(incompleteBlockNode), completedBlocks), _) =>
      ;     ; quick exit over the remaining parents
      ;     (Some(incompleteBlockNode), completedBlocks)
      ;   case ((None, completedBlocks), parentBlockNode) =>
      ;     ; okay, we haven't yet established that there are any blocks left to do...
      ;     val completed = parentBlockNode.children.count { blockNode => completedBlocks(blockNode.value.id) }
      ;     ; check if the parent block's limit ("quota") has been reached
      ;     val quota = blockNodeQuota(parentBlockNode)
      ;     if (completed >= quota) {
      ;       ; okay, we've completed all there is to complete in this block; mark it complete and keep going up
      ;       parentBlockNode match {
      ;         ; if it's not a pseudo-Node, like the root, that is
      ;         case BranchNode(_, value) =>
      ;           Response.create(participant_id, value.id)
      ;           (None, completedBlocks + value.id)
      ;         case RootNode(_) =>
      ;           (None, completedBlocks)
      ;       }
      ;     }
      ;     else {
      ;       ; welp, this is it, there's something left to do in this block
      ;       (Some(parentBlockNode), completedBlocks)
      ;     }
      ; }
      ; nextParentBlockNodeOption.map { nextParentBlockNode =>
      ;   nextPhysicalBlock(nextParentBlockNode, allCompletedBlocks).value
      ; }

(defn updateTree
  "Flattens the structured tree into a flat sequence of blocks.
  * `root` is the root node containing all blocks to be updated
  * `experiment_id` is the experiment ID"
  [root experiment_id]
  (let [allBlocks_original [] ; root.flatten.map(_.value)]
        ; 1. instantiate the missing blocks so that they have id's we can use when flattening
        allBlocks []
        all-blocks-ids (map :id allBlocks)] ; allBlocks_original.map { block =>
        ;   if (block.id == 0) {
        ;     Block.create(experiment_id, block.template_id, block.context,
        ;                  block.view_order, block.randomize, block.parent_block_id, block.quota)
        ;   }
        ;   else {
        ;     block
        ;   }
        ; }
    ; okay, all blocks should have .id fields now, ready to move on and fix the structure as needed
    ; TODO: set parent_block_id = null for rootBlocks
    ; TODO: set parent_block_id = the parent's id for the rest of the blocks (persist the tree structure)
    ;childBlock.parent_block_id = block.id
    ; now the tree structure is defined on each block, and the 'children' links are no longer needed.
    ;val allBlockIds = allBlocks.map(_.id)
    (log/info "Updating" (count allBlocks) "blocks")
    ; 3. brute-force: update them all
    (doseq [block allBlocks]
      ; careful, if block.id is null, this could wreak havoc and lose data!
      (update! (:id block) experiment_id (assoc block :experiment_id experiment_id)))
    ; 4. delete all the other blocks
    (db/delete! "block" ["experiment_id = ? AND NOT (id = ANY(?::integer[]))" experiment_id all-blocks-ids])))
